File: CoffeeMachine.java
package machine;

import java.util.Scanner;

public class CoffeeMachine {

    final static Scanner scanner = new Scanner(System.in);
    final static Machine coffeeMachine = new Machine();

    public static void main(String[] args) {
        boolean use_machine = true;
        while (use_machine) {
            System.out.println("Write action (buy, fill, take, remaining, exit):");
            String action = scanner.nextLine().trim().toLowerCase();
            switch (action) {
                case "buy" -> {
                    System.out.println("What do you want to buy? 1 - espresso, 2 - latte, 3 - cappuccino, back - to main menu:");
                    String option = scanner.nextLine().toLowerCase();
                    switch (option) {
                        case "1" -> coffeeMachine.makeCoffee(250, 0, 16, 4);
                        case "2" -> coffeeMachine.makeCoffee(350, 75, 20, 7);
                        case "3" -> coffeeMachine.makeCoffee(200, 100, 12, 6);
                        case "back" -> {}
                        default -> System.out.println("Sorry, unknown command");
                    }
                }
                case "fill" -> {
                    System.out.println("Write how many ml of water you want to add:");
                    int waterToAdd = scanner.nextInt();
                    System.out.println("Write how many ml of milk you want to add:");
                    int milkToAdd = scanner.nextInt();
                    System.out.println("Write how many grams of coffee beans you want to add:");
                    int coffeeToAdd = scanner.nextInt();
                    System.out.println("Write how many disposable cups of coffee you want to add:");
                    int cupsToAdd = scanner.nextInt();
                    coffeeMachine.fillMachine(waterToAdd, milkToAdd, coffeeToAdd, cupsToAdd);
                }
                case "take" -> coffeeMachine.takeMoney();
                case "remaining" -> coffeeMachine.printMachineStatus();
                case "exit" -> use_machine = false;
                default -> System.out.println("Sorry, unknown action");
            }
        }
    }
}

class Machine {
    int moneyInMachine = 550;
    int waterInMachine = 400;
    int milkInMachine = 540;
    int coffeeInMachine = 120;
    int cupsInMachine = 9;


    public void printMachineStatus() {
        System.out.println("The coffee machine has:");
        System.out.printf("%d ml of water \n", waterInMachine);
        System.out.printf("%d ml of milk \n", milkInMachine);
        System.out.printf("%d g of coffee beans \n", coffeeInMachine);
        System.out.printf("%d disposable cups \n", cupsInMachine);
        System.out.printf("$%d of money \n", moneyInMachine);
    }
    public void makeCoffee(int waterNeeded, int milkNeeded, int coffeeNeeded, int money) {
        if (checkMachineResources(waterNeeded, milkNeeded, coffeeNeeded)) {
            waterInMachine -= waterNeeded;
            milkInMachine -= milkNeeded;
            coffeeInMachine -= coffeeNeeded;
            cupsInMachine -= 1;
            moneyInMachine += money;
            System.out.println("I have enough resources, making you a coffee!");
        }
    }

    public boolean checkMachineResources(int water, int milk, int coffee) {
        if (waterInMachine < water) { System.out.println("Sorry, not enough water!"); return false; }
        if (milkInMachine < milk) { System.out.println("Sorry, not enough milk!"); return false; }
        if (coffeeInMachine < coffee) { System.out.println("Sorry, not enough coffee beans!"); return false; }
        if (cupsInMachine < 1) { System.out.println("Sorry, not enough cups!"); return false; }
        return true;
    }
    public void fillMachine(int waterRefill, int milkRefill, int coffeeRefill, int cupsRefill) {
        waterInMachine += waterRefill;
        milkInMachine += milkRefill;
        coffeeInMachine += coffeeRefill;
        cupsInMachine += cupsRefill;
    }

    public void takeMoney() {
        System.out.printf("I gave you $%d \n", moneyInMachine);
        moneyInMachine = 0;
    }

}

File: CoffeeMachine.kt
package machine

val ESPRESSO_RECIPE: Map<String, Int> = mapOf("waterPerCup" to 250, "milkPerCup" to 0, "beansPerCup"
        to 16, "cost" to 4)
val LATTE_RECIPE: Map<String, Int> = mapOf("waterPerCup" to 350, "milkPerCup" to 75, "beansPerCup"
        to 20, "cost" to 7)
val CAPPUCCINO_RECIPE: Map<String, Int> = mapOf("waterPerCup" to 200, "milkPerCup" to 100, "beansPerCup"
        to 12, "cost" to 6)

class CoffeeMachine(private var water: Int = 400,
                    private var milk: Int = 540,
                    private var beans: Int = 120,
                    private var cups: Int = 9,
                    private var money: Int = 550
) {

    /* Functionality */

    /* Status printer */
    private fun printStatus() {
        println("""The coffee machine has:
            ${this.water} ml of water
            ${this.milk} ml of milk
            ${this.beans} g of coffee beans
            ${this.cups} disposable cups
            $${this.money} of money""".trimIndent()
        )
    }

    /* Refill */
    private fun fillMachine(waterRefill: Int, milkRefill: Int, beansRefill: Int, cupsRefill: Int) {
        this.water += waterRefill
        this.milk += milkRefill
        this.beans += beansRefill
        this.cups += cupsRefill
    }

    private fun fillMenu() {
        println("Write how many ml of water you want to add:")
        val waterInput: Int = readln().toInt()
        println("Write how many ml of milk you want to add:")
        val milkInput: Int = readln().toInt()
        println("Write how many grams of coffee beans you want to add:")
        val beansInput: Int = readln().toInt()
        println("Write how many disposable cups you want to add:")
        val cupsInput: Int = readln().toInt()

        this.fillMachine(waterInput, milkInput, beansInput, cupsInput)
    }

    /* Emptying cash */
    private fun takeMoney() {
        println("I gave you $${this.money}")
        this.money = 0
    }

    /* Sell Coffee */
    private fun coffeeMenu() {
        println("What do you want to buy? 1 - espresso, 2 - latte, 3 - cappuccino:, back")
        when (readln()) {
            "1" -> makeCoffee(
                ESPRESSO_RECIPE["waterPerCup"]!!, ESPRESSO_RECIPE["milkPerCup"]!!,
                ESPRESSO_RECIPE["beansPerCup"]!!, ESPRESSO_RECIPE["cost"]!!)
            "2" -> makeCoffee(LATTE_RECIPE["waterPerCup"]!!, LATTE_RECIPE["milkPerCup"]!!,
                LATTE_RECIPE["beansPerCup"]!!, LATTE_RECIPE["cost"]!!)
            "3" -> makeCoffee(CAPPUCCINO_RECIPE["waterPerCup"]!!, CAPPUCCINO_RECIPE["milkPerCup"]!!,
                CAPPUCCINO_RECIPE["beansPerCup"]!!, CAPPUCCINO_RECIPE["cost"]!!)
            "back" -> {}
            else -> println("Unknown recipe")
        }
    }

    private fun makeCoffee(water: Int, milk: Int, beans: Int, money: Int) {
        when {
            this.water < water -> println("Sorry, not enough water!")
            this.milk < milk -> println("Sorry, not enough milk!")
            this.beans < beans -> println("Sorry, not enough coffee beans")
            this.cups < 1 -> println("Sorry, not enough disposable cups")
            else -> {
                println("I have enough resources, making you a coffee!")
                this.water -= water
                this.milk -= milk
                this.beans -= beans
                this.cups -= 1
                this.money += money
            }
        }

    }

    /* Main Menu */
    fun mainMenu() {
        var userInput = ""
        while(userInput != "exit") {
            when(userInput) {
                "buy" -> coffeeMenu()
                "fill" -> fillMenu()
                "take" -> takeMoney()
                "remaining" -> printStatus()
                else -> {}
            }
            println("Write action (buy, fill, take, remaining, exit):")
            userInput = readln().lowercase()
        }
    }
}